exonum/src/helpers/merkle.rs

```diff
@@ -19,11 +19,11 @@ use crypto::{self, Hash, HashStream};
 /// Computes Merkle root hash for a given list of hashes.
 ///
 /// If `hashes` are empty then `Hash::zero()` value is returned.
-pub fn root_hash<'a, I: IntoIterator<Item = &'a Hash>>(hashes: I) -> Hash {
-    let mut current_hashes: Vec<Hash> = hashes.into_iter().cloned().collect();
-    if current_hashes.is_empty() {
+pub fn root_hash(hashes: &[Hash]) -> Hash {
+    if hashes.is_empty() {
         return Hash::zero();
     }
+    let mut current_hashes = hashes.to_vec();
```

Comment:
Why do you need to convert hashes to vec? It's an additional space allocation that could be easily avoided.

Reply:
We need to have some storage to keep intermediate hashes, and its quite readable. Is there any obvious way to compute the root hash without any intermediate storage, stacks, etc. that I am not aware of? I'm sorry, I can't quite see how we can avoid intermediate storage completely.

Reply:
I'm not talking about every intermediate storage, just about initial one. If you'll write 
```
let mut current_hashes = combine_hash_list(hashes);
```
then you'll avoid the first allocation and the first `Vec` will contain only `N/2` hashes.

Reply:
This won't be that straightforward with in-place implementation suggested by @slowli. It uses

```rust
fn combine_hash_list(hashes: &mut Vec<Hash>)
```

and avoids any additional `Vec<Hash>` allocations except for the initial one. (Currently we allocate a new Vec for each intermediate batch.)

I'm not sure that allocating half less storage is worth the readability hit as in that case we will need to effectively duplicate implementation of `combine_hash_list()`:

```rust
/// Initial version used for first computation and allocation.
fn combine_inital_hash_list(hashes: &[Hash]) -> Vec<Hash> {
    // current implementation
}

/// In-place version used for subsequent iterations.
fn combine_hash_list(hashes: &mut Vec<Hash>) {
    // in-place implementation which does effectively the same
}
```

Should we do it like that?